### ISSUE TITLE
ipv6 client connection fails Error: connect EINVAL

### ISSUE THREAD

**Issue by MasterJames** [theturtle32/WebSocket-Node#278]
I got the server listening with ipv6 and a percent % with the interface name following For local prefix fe80, but when I go to use Client to connect with wss://**[fe80::a00:29?f:fdc2:e0?c]**/path
There is no connection with `_Error: connect EINVAL_ `

If I include the Ethernet device name wss://[fe80::a00:380f:fe2f:67fd%enp0s3]/path 
```
Error: You must specify a full WebSocket URL, including hostname. Relative URLs are not supported.
    at WebSocketClient.connect 
```
while a browser is good with just `[i::p:v:6]` no devicename and curl wants and likes it with the device name?
Seems like somewhere in the Client connect it's improperly understanding proper acceptable ipv6 addresses?!

Note you can also send a %25 to represent the % sign character and again the same with curl it will understand the url encoded percent sign. I've tried these with it including trying to backslash the percent sign, so not too many more options left I can think of? Maybe I need something other then ipv6 local addresses prefixed with fe80? alas that would be pretty hacking. Another hack might involve DNS.

**Comment by MasterJames** [comment]
WebSocketClient line 132
`        this.url = url.parse(requestUrl);`

Wrong result...
```
Url {protocol: "wss:", slashes: true, auth: null, host: "", port: null…}
auth: null
hash: null
host: ""
hostname: ""
href: "wss:///[fe80::a00:27ff:fe2f:69ff%enp0s3]/srvlnk"
path: "/[fe80::a00:27ff:fe2f:69ff%enp0s3]/srvlnk"
pathname: "/[fe80::a00:27ff:fe2f:69ff%enp0s3]/srvlnk"
port: null
protocol: "wss:"
query: null
search: null
slashes: true
```
Bug confirmed. Please make this package usable in a modern IOT world, thank you?!

**Comment by ibc** [comment]
Hi, just to be clear:

* `fe80::2e0:4cff:fe68:23ff` is a valid IPv6.
* `[fe80::2e0:4cff:fe68:23ff]` is a valid IPv6 reference (which means it can be used within a HTTP/WS URI).
* `fe80::2e0:4cff:fe68:23ff%en7` is not a valid IPv6 at all. The `%` symbol is NOT a valid char within a IPv6.
* So `[fe80::2e0:4cff:fe68:23ff%en7]` is not a valid IPv6 reference.

So `url.parse()` is behaving correctly. As you can see, the behavior of `url.parse()` is still the same in Node 7.

To summarize:

* `The IPv6 grammar⁄syntax clearly states that `%` is not a valid symbol within a IPv6.
* If you attempt to use an IPv6 with `%` in Node it will fail (as long as any module calls `url.parse()` with it), so this is not an issue of `Node-WebSocket`.
* `WebSocket-Node` should not do a specific hack to allow this wrong IPv6.

**Comment by MasterJames** [comment]
Here is the patch that works for me nothing else would.
See https://github.com/theturtle32/WebSocket-Node/pull/279

**Comment by ibc** [comment]
> Here is the patch/PR that works for me nothing else would. See #279

My point is that such a workaround should not be done within Node-WebSocket. You should not pass any IPv6 with a `%` symbol and expect that a library binds a local socket on it. Because it's an invalid IPv6.

> Not sure where the problem really is as you've pointed out thank you for your time.

The problem is that it's not a valid IPv6. Nothing else.

**Comment by MasterJames** [comment]
Well it only works that way with your Server it won't bind to it because it's a local range and needs the device name appended with a percent to understand where to route it or something I think? Maybe the specs wrong then?

**Comment by ibc** [comment]
Then there is a definitive issue with IPv6 and local ranges. The IPv6 BNF grammar does not allow (AFAIK) the `%` symbol into it. May you point please to any spec/RFC indicating that?

**Comment by ibc** [comment]
BTW, in your #279: Instead of that, you may directly pass a URL object to the `connect()` method.

**Comment by MasterJames** [comment]
Here is something I mention in the PR something about that but that it seems more fool proof and easier, but it's a good point I should explore that.
https://superuser.com/questions/99746/why-is-there-a-percent-sign-in-the-ipv6-address
it points to the so called least difficult they could find here...
https://msdn.microsoft.com/en-us/library/aa917150.aspx

**Comment by ibc** [comment]
In [RFC 3986](https://tools.ietf.org/html/rfc3986): Uniform Resource Identifier (URI): Generic Syntax

```
   A host identified by an IPv6 literal address is represented inside
   the square brackets without a preceding version flag.  The ABNF
   provided here is a translation of the text definition of an IPv6
   literal address provided in [RFC3513].  This syntax does not support
   IPv6 scoped addressing zone identifiers.
```

So, to be clear, you should NOT include a IPv6 scope ID into a URL.

I understand your issue, but it's not something that should be fixed within Node-WebSocket.

**Comment by ibc** [comment]
Said that, it's true that calling `bind()` with a link-local IPv6 with scope Id does work.
